### PR TITLE
Detekt - Resolve/Suppress All Baseline Warnings - Long methods warnings_ HomePageSettingsDialog, AuthorUseCase, ClicksUseCase, ReferrersUseCase

### PR DESCRIPTION
**Pull request by Hoossayn**
Parent: https://github.com/wordpress-mobile/WordPress-Android/issues/17010

This PR resolves/suppresses all complexity related LongMethod warnings for the HomePageSettingsDialog class, AuthorUseCase class, ClicksUseCase class, and ReferrersUseCase class (see docs [here](https://detekt.dev/docs/rules/complexity#longmethod)):

`4` x LongMethod  (Resolve: ad6bb237a2fb36f796156f9bede0bffd9a66a846 + 1cbf13249099f737bde8e23c46b9dff45fb4f1ff + f887ee07a8cb37ed37dc3b7e545ea717ca435e9e + 4e0cee3cf46405215c5b3cc9c98fdab97bc3bb84 + ccbb810665b27e5e88b2f0f66872046d02ff87e1 + 0431cde47a8ea2377cd69e177b31f66833bf6ea8 + 55311df61850b15373180b119e20fe5d5c065ec5 + 9432e4812c388ff9e87833b4e6eed9df350ab938 )

-----

To test:

- There is nothing much to test here.

- Verifying that all the CI checks are successful should be enough (especially the detekt check).
- However, if you really want to be thorough, you could smoke test the WordPress and/or Jetpack apps to verify that everything works as expected on every screen that relates to these changes. Specifically, you could follow the below steps:

   - Go to the menu tab
   - Click on site settings
   - Select HomePage settings
   - Test that you can change between different homepage settings 

To Test `Stats` 
   - Go to the menu tab
   - Click on Stat
   - Click on `Days` tab and confirm that the correct data for `Referrers`, `Clicks`, and `Authors` are displaying 
   - Do same for `weeks`, `months` and `year` tab 

## Regression Notes
1. Potential unintended areas of impact
     can't think of any 

2. What I did to test those areas of impact (or what existing automated tests I relied on)
   See `To test` section above.

3. What automated tests I added (or what prevented me from doing so)
     N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by Hoossayn** [comment]
👋🏾 @ParaskP7

**Review by ParaskP7 — CHANGES_REQUESTED** [review]
👋 @Hoossayn !

Thank you so much for your contribution to WPAndroid with this PR! 🥇 

I have reviewed and tested this PR as per the instructions, along with triggering CI on it via a [draft](https://github.com/wordpress-mobile/WordPress-Android/pull/17448) PR, which I just pushed on the main repo, everything works as expected, good job! 🌟 

-----

However, and apologies for making it once more a bit difficult for you to merge, but I think you rushed a bit with this PR, let me explain: ✍️ 

1. Warning (⚠️): You included two unrelated changes, one for the `Site Settings` screens and its `Home Page` settings, which you explicitly included test instruction and I verified that everything works as expect, but also for the `Stats` screen and its `Referrers`, `Clicks` and `Authors` stat related cards, which you also didn't included test instructions for.

FYI: This means that my confidence in you testing this change is lower than expected and I would now rather had this split into 2 PRs than just one, one for the `Site Settings` screen with the `Hope Page` settings change, and one for the `Stats` screen with the `Stats Cards` change.

No matter, I tested it on your behalf and everything seems to be working as expected. But, please, do update the testing instructions. 🙏 

2. Warnings (⚠️): For the use case related changes you choose to do a refactor that I would recommend against, that is to pass the `items` list as a parameter to your extracted methods (see `items: MutableList<BlockListItem>`). Doing so is dangerous (see [pass by reference vs. pass by value in Java](https://www.google.com/search?q=pass+by+reference+and+pass+by+value+in+Java)), but also makes the code harder to reason about as multiple functions are adding to this list of items.

Instead, I would urge you to consider this alternative instead, the below is for `ClicksUseCase` only, but applies to all 3 use cases:
```
    override fun buildUiModel(domainModel: ClicksModel, uiState: SelectedClicksGroup): List<BlockListItem> {
        val isBlockMode = useCaseMode == BLOCK
        return if (domainModel.groups.isEmpty()) {
            buildUiModelForNoGroup(isBlockMode)
        } else {
            buildUiModelForGroups(domainModel, uiState, isBlockMode)
        }
    }

    private fun buildUiModelForNoGroup(
        isBlockMode: Boolean
    ): List<BlockListItem> {
        val items = mutableListOf<BlockListItem>()
        if (isBlockMode) items.add(Title(R.string.stats_clicks))
        items.add(Empty(R.string.stats_no_data_for_period))
        return items
    }

    private fun buildUiModelForGroups(
        domainModel: ClicksModel,
        uiState: SelectedClicksGroup,
        isBlockMode: Boolean
    ): List<BlockListItem> {
        val items = mutableListOf<BlockListItem>()
        if (isBlockMode) items.add(Title(R.string.stats_clicks))
        val header = Header(R.string.stats_clicks_link_label, R.string.stats_clicks_label)
        items.add(header)
        domainModel.groups.forEachIndexed { index, group ->
            ...
        }

        if (useCaseMode == BLOCK && domainModel.hasMore) {
            ...
        }
        return items
    }
```
Is the diff clear to you, and more so, how an alternative like that is a better fit for such a refactor? Let me know if not and I can further elaborate to help you understand. 🤔

3. Suggestion (💡): Consider removing the extra two, `ComplexMethod:HomepageSettingsDialog` and `ComplexMethod:ReferrersUseCase` from the `baseline.xml` as well. They can be removed as well. 🎉

4. Minor (🔍): With this ad6bb237a2fb36f796156f9bede0bffd9a66a846 change you also removed the SAM constructor. I would recommend you don't mix such refactoring into one commit, exactly like you did with this 1cbf13249099f737bde8e23c46b9dff45fb4f1ff commit followed. You could have included both removals of SAM constructors in this 1cbf13249099f737bde8e23c46b9dff45fb4f1ff second commit of yours. PS: This tip is for next time.

-----

Let me know about all that above and we would then progress accordingly. 💯

**Comment by oooooo99999** [comment]
> 
> 
> 
> 
> <table>
>   <thead>
>     <tr>
>       <th width="50"></th>
>       <th width="100%" data-danger-table="true">Warnings</th>
>     </tr>
>   </thead>
>   <tbody><tr>
>       <td>:warning:</td>
>       <td>PR has more than 300 lines of code changing. Consider splitting into smaller PRs if possible.</td>
>     </tr>
>   </tbody>
> </table>
> 
> 
> 
> <p align="right">
>   Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a>
> </p>
>

**Comment by Hoossayn** [comment]
> 1. but also for the `Stats` screen and its `Referrers`, `Clicks` and `Authors` stat related cards, which you also didn't included test instructions for.

 `but also for the Stats screen and its Referrers, Clicks and Authors stat related cards, which you also didn't included test instructions for.`

it didn't skip my mind to test or include the test steps. i wasn't sure if i should include test for it. because up until this time, i've made changes directly on a fragment/activity. going forward i will ask for clarification before you review instead of going mute


`Is the diff clear to you, and more so, how an alternative like that is a better fit for such a refactor?`
   The difference is clear, Did some research on the topic as well  


`You could have included both removals of SAM constructors in this second commit of yours`
   i already mistakenly made the commit before realising i haven't removed the constructors. thinking about it now, i could have simply undo commit actually 

Thanks for your suggestion and also i haven't updated the commits in this doc. kindly let me know if i should or go ahead and split this PR further

**Comment by ParaskP7** [comment]
👋 @Hoossayn !

FYI: You could use the `> Xyz` markdown syntax of GitHub to reply to comments like that, just like I am replying to you, see below (also see [Quoting text](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-text)). I hope it will help you next time.

> it didn't skip my mind to test or include the test steps. i wasn't sure if i should include test for it. because up until this time, i've made changes directly on a fragment/activity. going forward i will ask for clarification before you review instead of going mute

I see now. 👍 

As a general rule, whatever code is being touched, no matter if unit tests are included, it is better to manual test the change as well. For sure, it should be included in the PR description so that the reviewers know were to focus on during their testing.

Please don't feel that you need to ask for clarification next time, you could safely assume that we always manual test any of our changes, especially since our codebase in not fully covered with UI tests, but only some E2E tests.

Let me know if you still have doubts about that. 🙏 

> The difference is clear, Did some research on the topic as well

This is awesome that you did your own research, thank you for that, much appreciated! 🙇 

> i already mistakenly made the commit before realising i haven't removed the constructors. thinking about it now, i could have simply undo commit actually

Let me take a look at that and I'll get back to you on it. 👍 

> Thanks for your suggestion and also i haven't updated the commits in this doc. kindly let me know if i should or go ahead and split this PR further

Cool, let me take a look first and I'll let you know my recommendation with how to proceed forward with this PR. 💯

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/AuthorsUseCase.kt [hunk 1]
```diff
@@ -98,74 +98,98 @@ class AuthorsUseCase constructor(
     }
 
     override fun buildUiModel(domainModel: AuthorsModel, uiState: SelectedAuthor): List<BlockListItem> {
+        val isBlockMode = useCaseMode == BLOCK
+        return if(domainModel.authors.isEmpty()){
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/AuthorsUseCase.kt:102]: Warnings (⚠️): Consider updating the format of this `if/else` statement as it missing empty spaces between `if` and `else`.

PS: The same apply to the other 2 use cases.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/AuthorsUseCase.kt [hunk 2]
```diff
@@ -98,74 +98,98 @@ class AuthorsUseCase constructor(
...
-                                            R.string.stats_post_views_label,
-                                            post.views
-                                    )
-                            )
-                        })
-                        items.add(Divider)
-                    }
-                }
-            }
+        if (useCaseMode == BLOCK && domainModel.hasMore) {
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/AuthorsUseCase.kt:149]: Suggestion (💡): Now that the `isBlockMode` is extracted and used as a parameter to this function, you could reuse that and replace `useCaseMode == BLOCK` with it.

PS: Same apply for the `ClicksUseCase` use case.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/AuthorsUseCase.kt [hunk 3]
```diff
@@ -98,74 +98,98 @@ class AuthorsUseCase constructor(
...
+        }
+        return items
+    }
 
-            if (useCaseMode == BLOCK && domainModel.hasMore) {
-                items.add(
-                        Link(
-                                text = R.string.stats_insights_view_more,
-                                navigateAction = create(statsGranularity, this::onViewMoreClicked)
+
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/AuthorsUseCase.kt:160]: Minor (🔍): Consider removing the extra empty line here.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCase.kt [hunk 4]
```diff
@@ -111,106 +111,151 @@ class ReferrersUseCase(
     }
 
     override fun buildUiModel(domainModel: ReferrersModel, uiState: SelectedGroup): List<BlockListItem> {
+        val isViewAllMode = useCaseMode != VIEW_ALL
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCase.kt:114]: Warning (⚠️): Consider renaming that to `isNotViewAllMode` so as for it to not becoming misleading going forward.

PS: The same renaming should be done on the `buildUiModelForNoGroup(...)` and `buildUiModelForGroups(...)` levels and everywhere within this class that this naming is used.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCase.kt [hunk 5]
```diff
@@ -111,106 +111,151 @@ class ReferrersUseCase(
...
-        if (useCaseMode != VIEW_ALL) {
-            items.add(Title(R.string.stats_referrers))
-        }
+    private fun buildUiModelForGroups(
+        domainModel: ReferrersModel,
+        uiState: SelectedGroup,
+        isViewAllMode: Boolean
+    ): List<BlockListItem> {
+        val items = mutableListOf<BlockListItem>()
+        if (isViewAllMode) items.add(Title(R.string.stats_clicks))
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCase.kt:137]: Warning (⚠️): The usage of `R.string.stats_clicks` is most probably a copy-paste mistake from `ClicksUseCase` and should be corrected to `R.string.stats_referrers`.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCase.kt [hunk 6]
```diff
@@ -111,106 +111,151 @@ class ReferrersUseCase(
...
+        val header = Header(R.string.stats_referrer_label, R.string.stats_referrer_views_label)
+        buildPieChart(domainModel)
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCase.kt:139]: Blocker (🚫): You will notice that by doing this change you are effectively ignoring adding the returned items to the main lists of such items. As such you should think about how to best progress with this change and why this is blocking. Let me know and I can further help you with that.

FYI: I would recommend the following alternative:
1. First change the `buildPieChart(...)` to this simpler version of it as you don't need to return a list of items anyway:
```
    private fun buildPieChart(
        domainModel: ReferrersModel
    ): PieChartItem? {
        return if (BuildConfig.IS_JETPACK_APP && useCaseMode == BLOCK_DETAIL) {
            buildPieChartItem(domainModel)
        } else {
            null
        }
    }
```
2. Then check for this item's availability and only add it to the list when it is not null, see below:
```
buildPieChart(domainModel)?.let { items.add(it) }
```

PS: The same apply to the `showReferrer(...)` and `showMore(...)` functions that you extracted. All 3 of them are actually ignoring the returned list of items.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCase.kt [hunk 7]
```diff
@@ -111,106 +111,151 @@ class ReferrersUseCase(
...
+                items.add(ExpandableItem(headerItem, isExpanded) { changedExpandedState ->
+                    onUiState(SelectedGroup(if (changedExpandedState) group.groupId else null))
+                })
+                if (isExpanded) {
+                    showReferrer(group, header)
+                }
             }
         }
+
+        if (useCaseMode == VIEW_ALL && domainModel.hasMore) {
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCase.kt:189]: Blocker (🚫): This seems to be another mistake, where did you get this `if (useCaseMode == VIEW_ALL && domainModel.hasMore)` from? 🤔

FYI: I am only aware of the below logic, which is anyway already included within the extracted `showMore(...)` function:
```
val shouldShowViewMore = itemCount < domainModel.groups.size ||
                    (useCaseMode == BLOCK && domainModel.hasMore)
```

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCase.kt [hunk 8]
```diff
@@ -111,106 +111,151 @@ class ReferrersUseCase(
...
+                    },
+                    contentDescription = contentDescriptionHelper.buildContentDescription(
+                            header,
+                            referrer.name,
+                            referrer.views
+                    )
+            )
+        })
+        items.add(Divider)
+        return  items
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCase.kt:256]: Minor (🔍): Consider removing the extra space between the `return` and `items`.

**Review by ParaskP7 — CHANGES_REQUESTED** [review]
👋 @Hoossayn !

I did another pass on your latest changes and I recommend you take a closer look at my included comments below, especially the warning (⚠️) and blocker (🚫) related ones.

Then, and because this PR is already getting a bit out of hand, with 2 separate changes in terms of testing the business logic, maybe it is indeed better for you to redo it, a bit more carefully this time, and in isolation to each other, that is:
1. The 1st PR should only focus on `HomepageSettingsDialog`
2. The 2nd PR should only focus on these 3 `Stats` related use cases.

-----

FYI: In addition to the above, and fact that the use case related unit tests didn't fail on your changes, especially on the wrongly introduced changes for the `ReferrersUseCase`, I would highly recommend for you to take this opportunity and add unit tests, on each use case, to try to achieve as high of a coverage as possible first before going forward with any method extraction, that is, main code change. This is so that to verify that your change ends up resulting in the expected outcome.

Wdyt? 🤔

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCase.kt (hunk 6 above)
Comment (Hoossayn) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCase.kt:139]: @ParaskP7  did you mean `showReferrer(...)` and `showmore(...)` should also return `PieChartItem`

Comment (ParaskP7) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/granular/usecases/ReferrersUseCase.kt:139]: 👋 @Hoossayn !

Actually no, what I meant by that is that the same `buildPieChart(...)` problem applies to `showReferrer(...)` and `showmore(...)` extractions you did, but the solution to those are different, in a sense thy need to return a `list` or `item`, specific to their case, not necessary a `PieChartItem` item.

Does the above clear things for you? 🙏

**Comment by Hoossayn** [comment]
> I would highly recommend for you to take this opportunity and add unit tests, on each use case, to try to achieve as high of a coverage as possible first before going forward with any method extraction, that is, main code change. This is so that to verify that your change ends up resulting in the expected outcome

@ParaskP7 can you kindly give hint of what to test for in these cases. also i might have to close the draft PR i already have, as i went ahead to make the code changes before writing the test

**Comment by ParaskP7** [comment]
👋 @Hoossayn !

> @ParaskP7 can you kindly give hint of what to test for in these cases. also i might have to close the draft PR i already have, as i went ahead to make the code changes before writing the test

Sure, so, each of these use-cases has this `buildUiModel(...)` function that you are targeting to refactor, which returns a `List<BlockListItem>`. These functions have a lot of logic within them, lots of `if/else` and `items.add(...)`. We need to first create tests that assert that each of those paths are tested, the returns list of items is the expected one and make sure the coverage for `buildUiModel(...)` is to 100%. Then, we can start refactoring, and with every refactor we do, we should run all these newly created tests again to verify that nothing is broken.

Does that make sense? 🤔

**Comment by ParaskP7** [comment]
👋 @Hoossayn , I hope I find you well!

FYI: I am going to close this PR as it hasn't been worked on for a while and as such it is now a bit outdated.

PS: Feel free to open another PR at your convenience.

**Pull request closed without merging**